File: ocaml/problem08.ml
(*
let compress list = 
let rec aux acc p = function
| [] -> acc
| head :: rest when acc = [] -> aux [head] head rest
| head :: rest -> if head = p then ( aux acc p rest )
                              else ( aux ( head :: acc ) head rest ) in
List.rev ( aux [] "x" list );;
*)

let rec compress = function
| a :: ( b :: _ as t ) -> if a = b then compress t else a :: compress t
| smaller -> smaller;;

compress ["a";"a";"a";"a";"b";"c";"c";"a";"a";"d";"e";"e";"e";"e"];;

File: haskell/problem08.hs
aux a b
  | b == [] = [a]
  | a == head b = b
  | otherwise = a:b

compress:: String -> String
compress x = foldr aux "" x

compress2:: String -> String
compress2 x = let aux2 a b
 			| b == [] = [a]
 			| a == head b = b
 			| otherwise = a:b
              in foldr aux2 "" x

compress3:: String -> String
compress3 x = foldr aux2 "" x
	      where aux2 a b = case b of
				[] -> [a]
				h:t  -> if h == a then b  else a:b

